keyvi/include/keyvi/dictionary/fuzzy_match.h

```diff
@@ -0,0 +1,131 @@
+      : fsa_(std::move(fsa)), query_(query), max_edit_distance_(max_edit_distance) {
+    std::vector<int> codepoints;
+    utf8::unchecked::utf8to32(query.begin(), query.end(), back_inserter(codepoints));
+
+    query_char_length_ = codepoints.size();
+    metric_ptr_.reset(new stringdistance::Levenshtein(codepoints, 20, max_edit_distance_));
+
+    const size_t minimum_exact_prefix_in_chars = 2;
+    size_t exact_prefix_in_bytes = 0;
+    while (exact_prefix_in_chars_ < minimum_exact_prefix_in_chars && exact_prefix_in_bytes < query_.size()) {
+      exact_prefix_in_bytes += util::Utf8Utils::GetCharLength(query[exact_prefix_in_bytes]);
+      ++exact_prefix_in_chars_;
+    }
+
+    // match exact prefix
+    uint64_t state = fsa_->GetStartState();
+    size_t depth = 0;
+    size_t utf8_depth = 0;
+    while (state != 0 && depth < exact_prefix_in_chars_) {
+      const size_t code_point_length = util::Utf8Utils::GetCharLength(query[utf8_depth]);
+      for (size_t i = 0; i < code_point_length; ++i, ++utf8_depth) {
+        state = fsa_->TryWalkTransition(state, query[utf8_depth]);
+        if (0 == state) {
+          break;
+        }
+      }
+      metric_ptr_->Put(codepoints[depth], depth);
+      ++depth;
+    }
+
+    if (state) {
+      if (depth == query_char_length_ && fsa_->IsFinalState(state)) {
+        first_match_ = Match(0, query_char_length_, query, 0, fsa_, fsa_->GetStateValue(state));
+      }
+      traverser_ptr_.reset(new fsa::CodePointStateTraverser<fsa::WeightedStateTraverser>(fsa_, state));
+    }
+  }
+
+  Match FirstMatch() const { return first_match_; }
+
+  Match NextMatch() {
+    for (; traverser_ptr_ && *traverser_ptr_; (*traverser_ptr_)++) {
+      const int intermediate_score = metric_ptr_->Put(traverser_ptr_->GetStateLabel(), candidate_length() - 1);
+      // don't consider subtrees which can not be matched anyways
+      if (query_char_length_ > candidate_length() && intermediate_score > max_edit_distance_) {
+        traverser_ptr_->Prune();
+        continue;
+      }
+
+      if (query_char_length_ + max_edit_distance_ < candidate_length()) {
+        traverser_ptr_->Prune();
+        continue;
+      }
+
+      if (traverser_ptr_->IsFinalState() && metric_ptr_->GetScore() <= max_edit_distance_) {
+        Match m = Match(0, candidate_length(), metric_ptr_->GetCandidate(), metric_ptr_->GetScore(),
+                        traverser_ptr_->GetFsa(), traverser_ptr_->GetStateValue());
+        (*traverser_ptr_)++;
+        return m;
+      }
+    }
+    return Match();
+  }
+
+ private:
+  size_t candidate_length() const {
+    assert(traverser_ptr_);
+    return exact_prefix_in_chars_ + traverser_ptr_->GetDepth();
+  }
+
+ private:
+  fsa::automata_t fsa_;
+  std::string query_;
+  const size_t max_edit_distance_;
+
+  size_t exact_prefix_in_chars_ = 0;
+  size_t query_char_length_ = 0;
+
+  Match first_match_;
+
+  std::unique_ptr<stringdistance::Levenshtein> metric_ptr_;
```

Comment:
that looks quite nice, so it would be easy to turn that into 

template<class MetricT = stringdistance::Levenshtein>
class FuzzyMatch final { 
...
};

Not saying we need that now.

Reply:
Sure thing!